staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller_debug.go

```diff
@@ -0,0 +1,199 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flowcontrol
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"text/tabwriter"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apiserver/pkg/server/mux"
+)
+
+func (cfgCtl *configController) Install(c *mux.PathRecorderMux) {
+	// TODO(yue9944882): handle "Accept" header properly
+	// debugging dumps a CSV content for three levels of granularity
+	// 1. row per priority-level
+	c.UnlistedHandleFunc("/debug/flowcontrol/dump_per_prioritylevel", cfgCtl.dumpPerPriorityLevel)
+	// 2. row per queue
+	c.UnlistedHandleFunc("/debug/flowcontrol/dump_per_queue", cfgCtl.dumpPerQueue)
+	// 3. row per request
+	c.UnlistedHandleFunc("/debug/flowcontrol/dump_per_request", cfgCtl.dumpPerRequest)
+}
+
+func (cfgCtl *configController) dumpPerPriorityLevel(w http.ResponseWriter, r *http.Request) {
+	cfgCtl.lock.Lock()
+	defer cfgCtl.lock.Unlock()
+	tabWriter := tabwriter.NewWriter(w, 8, 0, 1, ' ', 0)
+	columnHeaders := []string{
+		"PriorityLevelName",
+		"FlowSchemaName",
+		"ActiveQueues",
+		"IsIdle",
+		"RobinIndex",
+		"WaitingRequests",
+		"ExecutingRequests",
+	}
+	tabPrint(tabWriter, rowForHeaders(columnHeaders))
+	for _, plState := range cfgCtl.priorityLevelStates {
+		if plState.queues == nil {
+			continue
+		}
+		queueSetDigest := plState.queues.Dump()
```

Comment:
Is this producing anything worth the cost?  I am not convinced.  I think dumpPerPriorityLevel should be as lightweight as possible, since it is at that end of our spectrum.

Reply:
given the fact that we're locking the entire config controller on all granularities of dumping, there's always a significant cost at apiserver. i'm not sure if making this "lightweight" would help in any cases..

Reply:
Since it is locking the entire config controller --- and thus the request filter --- it is worthwhile to minimize the amount of time that the lock is held.